### PR TITLE
Skip doctest collection for subpackages depending on available dependencies

### PR DESCRIPTION
**Pull request by astrofrog**
This is an experiment to try out the proposed fix for https://github.com/astropy/pytest-doctestplus/pull/112 which allows doctest collection to be skipped per subpackage based on requirements.

The first commit's CI should fail before I actually define any options in ``setup.cfg`` (this is a deliberate test). Once I've checked that it fails, I'll push the fixes.

### CODE REVIEW

**Comment by astrofrog** [comment]
Ok so that fails nicely. Now pushing the skipping configuration!

**Comment by astrofrog** [comment]
Ok so let's see if that works. Note to self that I should find a way to avoid the following warning during collection though:

```
collecting ... The 'asdf' distribution was not found and is required by the application
collecting 8561 items / 22 skipped / 8539 selected
```

**Comment by astrofrog** [comment]
It works!

File: setup.cfg [hunk 1]
```diff
@@ -114,6 +114,10 @@ filterwarnings =
     ignore:PY_SSIZE_T_CLEAN will be required for '#' formats
     ignore:::astropy.tests.plugins.display
     ignore:::astropy.tests.disable_internet
+doctest_norecursedirs =
+    */setup_package.py
+doctest_subpackage_requires =
+    astropy/io/misc/asdf/*=asdf
```
Comment (pllim) [inline setup.cfg:120]: Spaces around `=` is even more consistent! 😉 
```suggestion
    astropy/io/misc/asdf/* = asdf
```

**Comment by astrofrog** [comment]
> Hmm, doesn't appear we have a job that runs without optional dependencies here but that is out of scope (see #10342).

Good point, I'll add a job to the Travis matrix with minimal dependencies, in the initial stage of testing

**Comment by astrofrog** [comment]
@pllim - to test this more, I've also (for now) reverted some of the changes in https://github.com/astropy/astropy/pull/10257 since we can now more easily skip the doctest collection, and I think the current structure is more convoluted with checks and exceptions all the time. I think it's reasonable to expect that importing astropy.io.misc.yaml should fail if PyYAML is not installed.

**Comment by astrofrog** [comment]
(still experimental btw, do not merge!)

**Comment by pllim** [comment]
Re: `yaml` -- Yes, sound fine. Thanks for cleaning up my mess!

**Comment by mhvk** [comment]
Belated :+1: to this change - very nice to have the `yaml.py` file cleaned up (even though it was made messy only out of necessity).

**Pull request merged**